README.md
README
================
Helmut Schütz

  - [PowerTOST](#powertost)
      - [Supported Designs](#supported-designs)
      - [Purpose](#purpose)
      - [Supported](#supported)
          - [Power and Sample Size](#power-and-sample-size)
          - [Methods](#methods)
          - [Helpers](#helpers)
      - [Defaults](#defaults)
          - [Average Bioequivalence](#average-bioequivalence)
          - [Reference-Scaled Average
            Bioequivalence](#reference-scaled-average-bioequivalence)
          - [Dose-Proportionality](#dose-proportionality)
          - [Power Analysis](#power-analysis)
      - [Examples](#examples)
          - [Parallel Design](#parallel-design)
          - [Crossover Design](#crossover-design)
          - [Replicate Designs](#replicate-designs)
          - [Dose-Proportionality](#dose-proportionality-1)
          - [Power Analysis](#power-analysis-1)
          - [Speed Comparisons](#speed-comparisons)
      - [Installation](#installation)

<!-- README.md is generated from README.Rmd. Please edit that file -->

# PowerTOST

The package contains functions to calculate power and estimate sample
size for various study designs used in (not only bio-) equivalence
studies.  
Built 2019-08-25 with R 3.6.1.

## Supported Designs

    #>    design                        name    df
    #>  parallel           2 parallel groups   n-2
    #>       2x2               2x2 crossover   n-2
    #>     2x2x2             2x2x2 crossover   n-2
    #>       3x3               3x3 crossover 2*n-4
    #>     3x6x3             3x6x3 crossover 2*n-4
    #>       4x4               4x4 crossover 3*n-6
    #>     2x2x3   2x2x3 replicate crossover 2*n-3
    #>     2x2x4   2x2x4 replicate crossover 3*n-4
    #>     2x4x4   2x4x4 replicate crossover 3*n-4
    #>     2x3x3   partial replicate (2x3x3) 2*n-3
    #>     2x4x2            Balaam's (2x4x2)   n-2
    #>    2x2x2r Liu's 2x2x2 repeated x-over 3*n-2
    #>    paired                paired means   n-1

Although some replicate designs are more ‘popular’ than others, sample
size estimations are valid for *all* of the following designs:

| design  |  type   | sequences         | periods |
| :-----: | :-----: | ----------------- | :-----: |
| `2x2x4` |  full   | 2 `TRTR\|RTRT`    |    4    |
| `2x2x4` |  full   | 2 `TRRT\|RTTR`    |    4    |
| `2x2x4` |  full   | 2 `TTRR\|RRTT`    |    4    |
| `2x2x3` |  full   | 2 `TRT\|RTR`      |    3    |
| `2x2x3` |  full   | 2 `TRR\|RTT`      |    3    |
| `2x3x3` | partial | 3 `TRR\|RTR\|RRT` |    3    |

Whilst "2x4x4" four period full replicate designs with four sequences
(TRTR|RTRT|TRRT|RTTR *or* TRRT|RTTR|TTRR|RRTT) are supported, they
should be avoided due to confounded effects.

<small>[↑ TOC](#readme)</small>

## Purpose

For various methods power can be *calculated* based on

  - nominal *α*, coefficient of variation (*CV*), deviation of test from
    reference (*θ*<sub>0</sub>), acceptance limits {*θ*<sub>1</sub>,
    *θ*<sub>2</sub>}, sample size (*n*), and design.

For all methods the sample size can be *estimated* based on

  - nominal *α*, coefficient of variation (*CV*), deviation of test from
    reference (*θ*<sub>0</sub>), acceptance limits {*θ*<sub>1</sub>,
    *θ*<sub>2</sub>}, target power, and design.

<small>[↑ TOC](#readme)</small>

## Supported

### Power and Sample Size

Power covers balanced as well as unbalanced sequences in crossover or
replicate designs and equal/unequal group sizes in two-group parallel
designs. Sample sizes are always rounded up to achieve balanced
sequences or equal group sizes.

  - Average Bioequivalence (with arbitrary *fixed* limits).
  - Two simultaneous TOST procedures.
  - Non-inferiority *t*-test.
  - Ratio of two means with normally distributed data on the original
    scale based on Fieller’s (‘fiducial’) confidence interval.
  - ‘Expected’ power in case of uncertain (estimated) variability and/or
    uncertain *θ*<sub>0</sub>.
  - Reference-scaled bioequivalence based on simulations.
      - EMA: Average Bioequivalence with Expanding Limits (ABEL).  
      - FDA: Reference-scaled Average Bioequivalence (RSABE) for Highly
        Variable Drugs / Drug Products and Narrow Therapeutic Index
        Drugs (NTIDs).  
  - Iteratively adjust *α* to control the type I error in ABEL and
    RSABE.
  - Dose-Proportionality using the power model.

<small>[↑ TOC](#readme)</small>

### Methods

  - Exact
      - Owen’s Q.
      - Direct integration of the bivariate non-central
        *t*-distribution.
  - Approximations
      - Non-central *t*-distribution.
      - ‘Shifted’ central *t*-distribution.

<small>[↑ TOC](#readme)</small>

### Helpers

  - Calculate *CV* from *MSE* or *SE* (and vice versa).
  - Calculate *CV* from given confidence interval.
  - Calculate *CV<sub>wR</sub>* from the upper expanded limit of an ABEL
    study.
  - Confidence interval of *CV*.
  - Pool *CV* from several studies.
  - Confidence interval for given *α*, *CV*, point estimate, sample
    size, and design.
  - Calculate *CV<sub>wT</sub>* and *CV<sub>wR</sub>* from a (pooled)
    *CV<sub>w</sub>* assuming a ratio of intra-subject variances.
  - *p*-values of the TOST procedure.
  - Analysis tool for exploration/visualization of the impact of
    expected values (*CV*, *θ*<sub>0</sub>, reduced sample size due to
    dropouts) on power of BE decision.
  - Construct design matrices of incomplete block designs.

<small>[↑ TOC](#readme)</small>

## Defaults

  - *α* 0.05, {*θ*<sub>1</sub>, *θ*<sub>2</sub>} (0.80, 1.25). Details
    of the sample size search (and the regulatory settings in
    reference-scaled average bioequivalence) are printed.
  - Note: In all functions values have to be given as ratios, not in
    percent.

### Average Bioequivalence

*θ*<sub>0</sub> 0.95, target power 0.80, design "2x2" (TR|RT), exact
method (Owen’s Q).

### Reference-Scaled Average Bioequivalence

*α* 0.05, point estimate constraint (0.80, 1.25), homoscedasticity
(*CV<sub>wT</sup>* = *CV<sub>wR</sup>*), scaling is based on
*CV<sub>wR</sub>*, target power 0.80, design "2x3x3" (TRR|RTR|RRT),
approximation by the non-central *t*-distribution, 100,000 simulations.

  - EMA, WHO, Health Canada, and many others: Average bioequivalence
    with expanding limits (ABEL).
  - FDA: RSABE.

#### Highly Variable Drugs / Drug Products

*θ*<sub>0</sub> 0.90 as recommended by [Tóthfalusi and
Endrényi](https://ejournals.library.ualberta.ca/index.php/JPPS/article/download/11612/9489)
(2011).

###### EMA

Regulatory constant `0.76`, upper cap of scaling at *CV<sub>wR</sup>*
50%, evaluation by ANOVA.

###### Health Canada

Regulatory constant `0.76`, upper cap of scaling at *CV<sub>wR</sup>*
\~57.4%, evaluation by intra-subject contrasts.

###### FDA

Regulatory constant `log(1.25)/0.25`, linearized scaled ABE (Howe’s
approximation).

#### Narrow Therapeutic Index Drugs (FDA)

*θ*<sub>0</sub> 0.975, regulatory constant `log(1.11111)/0.1`, upper cap
of scaling at *CV<sub>wR</sup>* \~21.4%, design "2x2x4" (TRTR|RTRT),
linearized scaled ABE (Howe’s approximation), upper limit of the
confidence interval of *s<sub>wT</sup>*/*s<sub>wR</sup>* ≤2.5.

### Dose-Proportionality

*β*<sub>0</sub> (slope) `1+log(0.95)/log(rd)` where `rd` is the ratio of
the highest and lowest dose, target power 0.80, crossover design,
details of the sample size search suppressed.

### Power Analysis

Minimum acceptable power 0.70. *θ*<sub>0</sub>, design, conditions, and
sample size method depend on defaults of the respective approaches (ABE,
ABEL, RSABE, NTID).

<small>[↑ TOC](#readme)</small>

## Examples

Before running the examples attach the library.

``` r
library(PowerTOST)
```

If not noted otherwise, defaults are employed.

### Parallel Design

Power for total *CV* 0.35, *θ*<sub>0</sub> 0.95, group sizes 52 and 49,
design "parallel".

``` r
power.TOST(CV = 0.35, theta0 = 0.95, n = c(52, 49), design = "parallel")
#> [1] 0.8011186
```

### Crossover Design

Sample size for assumed intra-subject *CV* 0.20.

``` r
sampleN.TOST(CV = 0.20)
#> 
#> +++++++++++ Equivalence test - TOST +++++++++++
#>             Sample size estimation
#> -----------------------------------------------
#> Study design: 2x2 crossover 
#> log-transformed data (multiplicative model)
#> 
#> alpha = 0.05, target power = 0.8
#> BE margins = 0.8 ... 1.25 
#> True ratio = 0.95,  CV = 0.2
#> 
#> Sample size (total)
#>  n     power
#> 20   0.834680
```

Sample size for equivalence of the ratio of two means with normality on
original scale based on Fieller’s (‘fiducial’) confidence interval.
*CV<sub>w</sub>* 0.20, *CV<sub>b</sub>* 0.40.  
Note the default *α* 0.025 (95% CI) of this function because it is
intended for studies with clinical endpoints.

``` r
sampleN.RatioF(CV = 0.20, CVb = 0.40)
#> 
#> +++++++++++ Equivalence test - TOST +++++++++++
#>     based on Fieller's confidence interval
#>             Sample size estimation
#> -----------------------------------------------
#> Study design: 2x2 crossover
#> Ratio of means with normality on original scale
#> alpha = 0.025, target power = 0.8
#> BE margins = 0.8 ... 1.25 
#> True ratio = 0.95,  CVw = 0.2,  CVb = 0.4
#> 
#> Sample size
#>  n     power
#> 28   0.807774
```

<small>[↑ TOC](#readme)</small>

### Replicate Designs

#### ABE

Sample size for assumed intra-subject *CV* 0.45, *θ*<sub>0</sub> 0.90,
three period full replicate study "2x2x3" (TRT|RTR *or* TRR|RTT).

``` r
sampleN.TOST(CV = 0.45, theta0 = 0.90, design = "2x2x3")
#> 
#> +++++++++++ Equivalence test - TOST +++++++++++
#>             Sample size estimation
#> -----------------------------------------------
#> Study design: 2x2x3 (3 period full replicate) 
#> log-transformed data (multiplicative model)
#> 
#> alpha = 0.05, target power = 0.8
#> BE margins = 0.8 ... 1.25 
#> True ratio = 0.9,  CV = 0.45
#> 
#> Sample size (total)
#>  n     power
#> 124   0.800125
```

Note that the conventional model assumes homoscedasticity. For
heteroscedasticity we can ‘switch off’ all conditions of one of the
methods for reference-scaled ABE. We assume a σ<sup>2</sup> ratio of ⅔
(*i.e.*, T has a lower variability than R). Only relevant columns of the
data.frame shown.

``` r
reg <- reg_const("USER", r_const = NA, CVswitch = Inf,
                 CVcap = Inf, pe_constr = FALSE)
CV  <- CVp2CV(CV = 0.45, ratio = 2/3)
res <- sampleN.scABEL(CV=CV, design = "2x2x3", regulator = reg,
                      details = FALSE, print = FALSE)
print(res[c(3:4, 8:9)], digits = 5, row.names = FALSE)
#>    CVwT    CVwR Sample size Achieved power
#>  0.3987 0.49767         126         0.8052
```

Similar sample size because the pooled *CV* is still 0.45.

<small>[↑ TOC](#readme)</small>

#### ABEL

Sample size assuming homoscedasticity (*CV<sub>w</sub>* = 0.45).

``` r
sampleN.scABEL(CV = 0.45, details = TRUE)
#> 
#> +++++++++++ scaled (widened) ABEL +++++++++++
#>             Sample size estimation
#>    (simulation based on ANOVA evaluation)
#> ---------------------------------------------
#> Study design: 2x3x3 (partial replicate)
#> log-transformed data (multiplicative model)
#> 1e+05 studies for each step simulated.
#> 
#> alpha  = 0.05, target power = 0.8
#> CVw(T) = 0.45; CVw(R) = 0.45
#> True ratio = 0.9
#> ABE limits / PE constraint = 0.8 ... 1.25 
#> EMA regulatory settings
#> - CVswitch            = 0.3 
#> - cap on scABEL if CVw(R) > 0.5
#> - regulatory constant = 0.76 
#> - pe constraint applied
#> 
#> 
#> Sample size search
#>  n     power
#> 36   0.7755 
#> 39   0.8059
```

Iteratively adjust *α* to control the Type I Error ([Labes,
Schütz](https://doi.org/10.1007/s11095-016-2006-1)). Slight
heteroscedasticity (*CV<sub>wT</sub>* 0.30, *CV<sub>wR</sub>* 0.35),
four period full replicate "2x2x4" study, 30 subjects, balanced
sequences.

``` r
scABEL.ad(CV = c(0.30, 0.35), design = "2x2x4", n = 30)
#> +++++++++++ scaled (widened) ABEL ++++++++++++
#>          iteratively adjusted alpha
#>    (simulations based on ANOVA evaluation)
#> ----------------------------------------------
#> Study design: 2x2x4 (4 period full replicate)
#> log-transformed data (multiplicative model)
#> 1,000,000 studies in each iteration simulated.
#> 
#> CVwR 0.35, CVwT 0.3, n(i) 15|15 (N 30)
#> Nominal alpha                 : 0.05 
#> True ratio                    : 0.9000 
#> Regulatory settings           : EMA (ABEL)
#> Switching CVwR                : 0.3 
#> Regulatory constant           : 0.76 
#> Expanded limits               : 0.7723 ... 1.2948
#> Upper scaling cap             : CVwR > 0.5 
#> PE constraints                : 0.8000 ... 1.2500
#> Empiric TIE for alpha 0.0500  : 0.06651
#> Power for theta0 0.9000       : 0.814
#> Iteratively adjusted alpha    : 0.03540
#> Empiric TIE for adjusted alpha: 0.05000
#> Power for theta0 0.9000       : 0.771
```

With the nominal *α* 0.05 the Type I Error will be inflated (0.0665).
With the adjusted *α* 0.0354 (*i.e.*, the 92.92% confidence interval)
the TIE will be controlled, although with a slight loss in power
(decreases from 0.814 to 0.771).  
Consider `sampleN.scABEL.ad(CV = c(0.30, 0.35), design = "2x2x4")` to
estimate the sample size which both controls the TIE and maintains the
target power. In this example 34 subjects will be required.

<small>[↑ TOC](#readme)</small>

#### RSABE

#### HVD(P)s

Sample size for a four period full replicate "2x2x4" study (any of
TRTR|RTRT, TRRT|RTTR, TTRR|RRTT) assuming heteroscedasticity
(*CV<sub>wT</sub>* 0.40, *CV<sub>wR</sub>* 0.50). Details of the sample
size search suppressed.

``` r
sampleN.RSABE(CV = c(0.40, 0.50), design = "2x2x4", details = FALSE)
#> 
#> ++++++++ Reference scaled ABE crit. +++++++++
#>            Sample size estimation
#> ---------------------------------------------
#> Study design: 2x2x4 (4 period full replicate)
#> log-transformed data (multiplicative model)
#> 1e+05 studies for each step simulated.
#> 
#> alpha  = 0.05, target power = 0.8
#> CVw(T) = 0.4; CVw(R) = 0.5
#> True ratio = 0.9
#> ABE limits / PE constraints = 0.8 ... 1.25 
#> Regulatory settings: FDA 
#> 
#> Sample size
#>  n    power
#> 20   0.81509
```

<small>[↑ TOC](#readme)</small>

#### NTIDs

Sample size assuming heteroscedasticity (*CV<sub>w</sub>* 0.125,
σ<sup>2</sup> ratio 2.5, *i.e.*, T has a substantially higher
variability than R). TRTR|RTRT according to the [FDA’s
guidance](https://www.accessdata.fda.gov/drugsatfda_docs/psg/Warfarin_Sodium_tab_09218_RC12-12.pdf).
Assess additionally which one of the three components (scaled, ABE,
*s<sub>wT</sub>*/*s<sub>wR</sub>* ratio) drives the sample size.

``` r
CV <- signif(CVp2CV(CV = 0.125, ratio = 2.5), 4)
n  <- sampleN.NTIDFDA(CV = CV)[["Sample size"]]
#> 
#> +++++++++++ FDA method for NTIDs ++++++++++++
#>            Sample size estimation
#> ---------------------------------------------
#> Study design:  2x2x4 (TRTR|RTRT) 
#> log-transformed data (multiplicative model)
#> 1e+05 studies for each step simulated.
#> 
#> alpha  = 0.05, target power = 0.8
#> CVw(T) = 0.1497, CVw(R) = 0.09433
#> True ratio     = 0.975 
#> ABE limits     = 0.8 ... 1.25 
#> Implied scABEL = 0.9056 ... 1.1043 
#> Regulatory settings: FDA 
#> - Regulatory const. = 1.053605 
#> - 'CVcap'           = 0.2142 
#> 
#> Sample size search
#>  n     power
#> 28   0.665530 
#> 30   0.701440 
#> 32   0.734240 
#> 34   0.764500 
#> 36   0.792880 
#> 38   0.816080
suppressMessages(power.NTIDFDA(CV = CV, n = n, details = TRUE))
#>        p(BE)  p(BE-sABEc)    p(BE-ABE) p(BE-sratio) 
#>      0.81608      0.93848      1.00000      0.85794
```

The *s<sub>wT</sub>*/*s<sub>wR</sub>* component shows the lowest power
and hence, drives the sample size.  
Compare that with homoscedasticity (*CV<sub>wT</sub>* =
*CV<sub>wR</sub>* = 0.125):

``` r
CV <- 0.125
n  <- sampleN.NTIDFDA(CV = CV, details = FALSE)[["Sample size"]]
#> 
#> +++++++++++ FDA method for NTIDs ++++++++++++
#>            Sample size estimation
#> ---------------------------------------------
#> Study design:  2x2x4 (TRTR|RTRT) 
#> log-transformed data (multiplicative model)
#> 1e+05 studies for each step simulated.
#> 
#> alpha  = 0.05, target power = 0.8
#> CVw(T) = 0.125, CVw(R) = 0.125
#> True ratio     = 0.975 
#> ABE limits     = 0.8 ... 1.25 
#> Regulatory settings: FDA 
#> 
#> Sample size
#>  n     power
#> 16   0.822780
suppressMessages(power.NTIDFDA(CV = CV, n = n, details = TRUE))
#>        p(BE)  p(BE-sABEc)    p(BE-ABE) p(BE-sratio) 
#>      0.82278      0.84869      1.00000      0.95128
```

Here the scaled ABE component shows the lowest power and drives the
sample size, which is much lower than in the previous example.

<small>[↑ TOC](#readme)</small>

### Dose-Proportionality

*CV* 0.20, Doses 1, 2, and 8 units, *β*<sub>0</sub> 1, target power
0.90.

``` r
sampleN.dp(CV = 0.20, doses = c(1, 2, 8), beta0 = 1, targetpower = 0.90)
#> 
#> ++++ Dose proportionality study, power model ++++
#>             Sample size estimation
#> -------------------------------------------------
#> Study design: crossover (3x3 Latin square) 
#> alpha = 0.05, target power = 0.9
#> Equivalence margins of R(dnm) = 0.8 ... 1.25 
#> Doses = 1 2 8 
#> True slope = 1, CV = 0.2
#> Slope acceptance range = 0.89269 ... 1.1073 
#> 
#> Sample size (total)
#>  n     power
#> 18   0.915574
```

Note that the acceptance range of the slope depends on the ratio of the

NEWS
Version 1.4-7
2018-04-12 (dedicated to the birthday of Alfie Schuetz :-))
      - Bug corrected in functions using expected power
      - Function type1error.2TOST() no longer available since it suffers from
        insufficient precision to obtain the type 1 error (TIE) via simulations.
        Due to the intersection-union principle the TIE is always upper bounded
        to alpha by theory.
      - Function power.2TOST() now based on simulations to obtain the power 
        of 2 TOSTs (statistical flaw in 4-dimensional t-distribution approach).
      - sampleN.2TOST and man page adapted according to this change.
      - regulator="FDA" implemented in scABEL.ad().
      - New function power.RSABE2L.sds() which implements the 'exact' based
        method for RSABE (ncTOST) of the two Laszlos.
        Documentation included.

Version 1.4-6
Released 2017-08-17
      - Bug fixed in functions relying on simulations if nsims>1e7.
      - Functions CVwRfromU() / U2CVwR() to calculate CVwR from the upper expanded 
        limit of an ABEL study according to EMA's or Health Canada's rules.
      - Power and sample size for TOST: Argument alpha restricted to scalar.
        Internal functions now allow length = 2 (different alpha's
        for the two null hypotheses).
      - Man pages for non-inferiority again updated.
      - Deprecated argument dfCV in expected power functions now completely removed.
      - Data lazy loading (allows access by the name)

Version 1.4-5
Released 2017-05-19
      - Man pages for non-inferiority functions updated to include a
        description of the underlying hypotheses.
      - Remaining man pages updated where term 'Null (true) ratio' was
        mentioned instead of 'True ratio'.
      - Two new test scrips in /test subdirectory added: test_ABEL.R and
        test_RSABE which recalculate the sample size tables given in 
        Tothfalusi & Endrenyi "Sample Sizes for Designing Bioequivalence Studies 
        for Highly Variable Drugs". Contributed by Helmut.
      - power.scABEL.sdsims() now has an argument design_dta to specify the
        design via a data.frame. May be useful for considering missing data.
        Attention! This feature is experimental because the data.frame is not 
        checked at moment.
      - Bug in power.scABEL.sdsims() fixed which gave power=NA if alpha=0.
      - Internal change of coding in design helpers (repeated creation of 
        data.frame with design characteristics avoided).
      - Typo corrected in Expected_Power_for_TOST.pdf.
      - Updated functions scABEL.ad() and sampleN.scABEL.ad() to allow
        subject data simulations via power.scABEL.sdsims() if regulator = "EMA".
        Removed regulator = "ANVISA".
        Changed the order of sequences to be consistent with the other
        functions of PowerTOST.
      - Renamed the variable 'adj. alpha' in sampleN.scABEL.ad() to 'alpha.adj'
        for consistency with scABEL.ad(). Bug corrected if pre-specified
        alpha gives TIE <0.05.
      - Removed regulator = "ANVISA" from pa.scABE().
      - Small bug in the S3 method plot for class 'pwrA' corrected (text
        in the 4th box was truncated on top if saved as a PDF).
        Replaced text() by TeachingDemos::shadowtext() to enhance legibility
        (interfered with underlying grid).

Version 1.4-4
Released 2017-03-15
      - Administrative change for expected power: cubature::adaptIntegrate 
        replaced by cubature::hcubature to reflect change of function name within
        package cubature.
      - Added function power.scABEL.sdsims() to calculate power for the BE decision 
        via scaled (widened) BE acceptance limits (EMA recommended) based on 
        subject data simulations.
      - Low CV (cosmetic) correction in power.scABEL() introduced.
      - Removed default CV = 0.3 from scABEL.ad() and sampleN.scABEL.ad().
        Stops execution if CV not specified.
      - DOI for references added in many man pages. THX to Helmut.
      - Bens description of expected power added in /doc subdirectory
      - BE_power_sample_size_excerpt.pdf updated to reflect the changes in
        computation of OwensQ.
      - Use of removed functions power.scABEL2() and sampleN.scABEL2() in
        pa.scABE() corrected.

Version 1.4-3
Released 2016-11-01
      - Little bug in power.TOST() removed which caused power <0 for large degrees
        of freedom.
      - Functions for expected power of the TOST procedure and sample size based
        on expected power reworked to deal also with uncertainty of theta0 or
        dealing with both uncertainties of CV and theta0.
        Contributed mainly by Benjamin Lang.
      - Functions for expected power of non-inferiority testing and sample size
        based on that expected power reworked to deal also with uncertainty of
        theta0 or dealing with both uncertainties of CV and theta0.
        Contributed mainly by Benjamin Lang.
      - Deprecated functions power.scABEL2() and sampleN.scABEL2() now removed.
        regulator = "ANVISA" no longer allowed in the scaled ABEL functions.
      - OwensQ() simpler/faster implemented. Now based soleyly on numerical
        integrataion in combination with non-central t.
      - Owen's T-function, used in OwensQOwen() now based on algorithm AS76 and
        remarks to that algorithm to avoid numerical errors of the implementation
        via integrate(). 
      - Data.frame ct9.6.6 in data("data2x2x3") added which was missing a long
        time ago.
      - Little inconsistency resolved (by Helmut) in output of power analysis 
        functions w.r.t. ratio or percent. Thanks to user myttiri of BEBA forum
        for pointing out this inconsistency.
        
Version 1.4-2
Released 2016-07-14
      - Implementation_scaledABE_sims.pdf in \doc subdirectory updated to

R/Fieller.R
#------------------------------------------------------------------------------
# power / sample size of ratio of means with normality on original scale
# (based on Fieller's CI and Sasabuchi's test)
# Hauschke D., Kieser M., Diletti E. and Burke M.
# "Sample size determination for proving equivalence based on the ratio
#  of two means for normally distributed data"
# Stat. Med. 18(1) p93-105 (1999) 
# 
# Author: dlabes, started coding Jul 2010
#------------------------------------------------------------------------------

# the multivariate t- and normal distribution with arbitrary corr
# require(mvtnorm)

power.RatioF <- function(alpha=0.025, theta1=0.8, theta2, theta0=0.95, CV, CVb, 
                         n, design="2x2", setseed=TRUE)
{
  if (!(tolower(design) %in% c("2x2","parallel"))) {
    design <- "2x2"; warning("Design unknown. Will use 2x2 cross-over.\n",
        call.=FALSE, immediate.=TRUE)
  }
  if (missing(CV)) stop("CV must be given!")
  if (tolower(design)!="parallel"){
    if (missing(CVb)) stop("CVb(etween) must be given in case of cross-over!",
                           call.=FALSE)
  }
  if (missing(n))  stop("Number of subjects must be given!", call.=FALSE)
  if (n<=2) stop("Number of subjects must be >2!", call.=FALSE)
  
  if (missing(theta2)) theta2 <- 1/theta1
  
  # without set.seed() each run gives different power values
  # due to the implementation of pmvt() and pmvnorm(), 
  # especially if power is small
  if (setseed) set.seed(123456789)
  power <- .power.RatioF(alpha, theta1, theta2, theta0, CV, CVb, n, design)

  return(power)
}
# -------------------------------------------------------------------------
# power working horse without any input check
.power.RatioF <- function(alpha, theta1, theta2, theta0, CV, CVb, n, design)
{
  df <- n-2 #for both designs
  
  # non-centrality parms
  delta <- c(0,0)
  if (tolower(design)=="parallel"){
    delta[1] <- (theta0 - theta1)/CV/sqrt(2.*(1.0+theta1^2)/n)
    delta[2] <- (theta0 - theta2)/CV/sqrt(2.*(1.0+theta2^2)/n)
    # in case of theta0=theta1 or =theta2 and CV=0 NaN is produced
    # setting it to zero, is this correct?
    delta[is.nan(delta)] <- 0
    rho <- (1.0+theta1*theta2)/sqrt((1+theta1^2)*(1+theta2^2))
  } else {
    # cross-over
    CVw <- CV
    delta[1] <- (theta0 - theta1)/sqrt((CVw^2*(1.0+theta1^2) + 
                 CVb^2*(1.0-theta1)^2)/n)
    delta[2] <- (theta0 - theta2)/sqrt((CVw^2*(1.0+theta2^2) + 
                 CVb^2*(1.0-theta2)^2)/n)
    # in case of theta0=theta1 or =theta2 and CV=0 NaN is produced
    # setting it to zero, is this correct?
    delta[is.nan(delta)] <- 0
    
    rho <- (CVw^2*(1.0+theta1*theta2) + 
            CVb^2*(1.0+theta1*theta2-theta1-theta2)) /
            sqrt((CVw^2*(1.0+theta1^2) + CVb^2*(1.0-theta1)^2)*
                 (CVw^2*(1.0+theta2^2) + CVb^2*(1.0-theta2)^2))
  }
  # correlation
  corr <- diag(2)
  corr[1,2] <- corr[2,1] <- rho
  # upper limits of integration
  upper <- c(Inf,0)
  upper[2] <- -qt(1-alpha,df)

  # TODO: vectorized form of that
  if (df<4000){
    power <- pmvt(upper=upper, delta=delta, corr=corr, df=df, abseps=1e-9)
    upper[1] <- qt(1-alpha,df)
    power <- power - pmvt(upper=upper, delta=delta, corr=corr, df=df,
                          abseps=1e-9)
  } else {
    # large sample approx. multivariate normal distri.
    power <- pmvnorm(upper=upper, mean=delta, corr=corr)
    upper[1] <- qt(1-alpha,df)
    power <- power - pmvnorm(upper=upper, mean=delta, corr=corr)
  }
  
  # get rid of the attributes
  attributes(power) <- NULL
  # power<0 may happen due to numeric inaccuracies
  power <- ifelse(power<0, 0, power)
  return(power)
}
# ------------------------------------------------------------------------
# internal function: start of sample size search
.sampleN0.RatioF <- function(alpha=0.025, targetpower, theta1, theta2, theta0,
                             CV, CVb, design="2x2")
{
  # Hauschke D., Steinijans V. and Pigeot I.
  # "Bioequivalence studies in Drug Development"
  # Chapter 10.3.3
  # John Wiley & Sons, Chichester (2007)
  z1 <- qnorm(1-alpha)
  # value 0.04 corresponds roughly to log(0.96)
  # with lower values there can be many steps between 0.95 and 1
  if (abs(log(theta0))>0.04) z2 <- qnorm(targetpower) else {
    z2 <- qnorm(1-(1-targetpower)/2)
    theta0 <- 1
  }
  
  if (theta0<=theta1 | theta0>=theta2) stop("Ratio0 ",theta0,
      " must be between theta1/theta2 = ", theta1," / ",theta2," !")
  if (tolower(design)=="parallel"){
    n01 <- 2.0*(1+theta1^2)*(CV/(theta1-theta0))^2*(z1+z2)^2
    n02 <- 2.0*(1+theta2^2)*(CV/(theta2-theta0))^2*(z1+z2)^2
  } else {
  # cross-over
    CVw <- CV
    n01 <- (CVw^2*(1.0+theta1^2)+ CVb^2*(1.0-theta1)^2)*(z1+z2)^2 / 
           (theta1-theta0)^2
    n02 <- (CVw^2*(1.0+theta2^2)+CVb^2*(1.0-theta2)^2)*(z1+z2)^2 / 
           (theta2-theta0)^2
  } 
  n0 <- max(n01,n02)
  #make an even multiple of 2
  n0=ceiling(n0)
  n0 <- 2*trunc(n0/2)
  if (n0<4) n0 <- 4   # minimum sample size
  return(n0)
} # end function
# ---------------------------------------------------------------------------
# Sample size for a desired power of ratio of two means 
# with normality on original scale
# mainly intended for studies with clinical endpoints
# therefore 95% CIs had to be used and consequently alpha=0.025
sampleN.RatioF <- function(alpha=0.025, targetpower=0.8, theta1=0.8, theta2,
                           theta0=0.95, CV, CVb, design="2x2", print=TRUE, 
                           details=FALSE, imax=100, setseed=TRUE)
{
  if (!(tolower(design) %in% c("2x2","parallel"))) {
    design <- "2x2"
    warning("Design unknown. Will use 2x2 cross-over.",
            call.=FALSE, immediate.=TRUE)
  }
  if (missing(theta2)) theta2 <- 1/theta1
  if ( (theta0<=theta1) | (theta0>=theta2) ) {
    stop("True ratio ",theta0," not between margins ",theta1," / ",theta2,"!", 
         call.=FALSE)
  }
  
  if (missing(CV)) stop("Err: CV must be given!", call.=FALSE)
  
  if (tolower(design)!="parallel"){
    if (missing(CVb)) stop("CVb(etween) must be given in case of cross-over!",
                           call.=FALSE)
  } else CVb <- 0
  
  if (print) {
    cat("\n+++++++++++ Equivalence test - TOST +++++++++++\n")
    cat("    based on Fieller's confidence interval\n")
    cat("            Sample size estimation\n")
    cat("-----------------------------------------------\n")
    if (tolower(design=="parallel")) {
      cat("Study design: 2-group parallel\n")
    } else {
      cat("Study design: 2x2 cross-over\n")
    }
    cat("Ratio of means with normality on original scale\n")
    cat("alpha = ",alpha,", target power = ", targetpower,"\n", sep="")
    cat("BE margins =",theta1,"...", theta2,"\n")
    if (tolower(design)=="parallel"){
      cat("True ratio = ",theta0,",  CV = ",CV,"\n", sep="")
    } else {
      cat("True ratio = ",theta0,",  CVw = ",CV,",  CVb = ",CVb,
          "\n", sep="")
    }
  }
  # Start
  n   <- .sampleN0.RatioF(alpha, targetpower, theta1, theta2, theta0, CV, CVb, 
                           design)
  if (setseed) set.seed(123456789)
  pow <- .power.RatioF(alpha, theta1, theta2, theta0, CV, CVb, n, design)
  if (details) {
    cat("\nSample size search\n")
    cat(" n     power\n")
    # do not print first too high
    if (pow<=targetpower) cat( n," ", formatC(pow, digits=6, format="f"),"\n")
  }
  
  steps <- 2
  iter  <- 0  # emergency brake
  # starting with too high power should be rare
  while (pow>targetpower) {
    if (n<=4) { # min number
      if (print & iter==0) cat( n," ", formatC(pow, digits=6, format="f"),"\n")
      break
    }
    n    <- n - steps
    iter <- iter+1
    if (setseed) set.seed(123456789)
    pow  <- .power.RatioF(alpha, theta1, theta2, theta0, CV, CVb, n, design)
    if (details) cat( n," ", formatC(pow, digits=6, format="f"),"\n")
    if (iter>imax) break 
  }
  while (pow<targetpower) {
    n    <- n+steps
    iter <- iter+1
    if (setseed) set.seed(123456789)
    pow <- .power.RatioF(alpha, theta1, theta2, theta0, CV, CVb, n, design)
    if (details) cat( n," ", formatC(pow, digits=6, format="f"),"\n")
    if (iter>imax) break 
  }
  if (print && !details) {
    cat("\nSample size\n")
    cat(" n     power\n")
    cat( n," ", formatC(pow, digits=6, format="f"),"\n")
  }
  #return
  res <- data.frame(design=design, alpha=alpha, CV=CV,CVb=CVb, theta0=theta0, 
                    theta1=theta1, theta2=theta2, n=n, power=pow, 
                    targetpower=targetpower)
  names(res) <-c("Design","alpha","CV","CVb","theta0","theta1","theta2",
                 "Sample size", "Achieved power", "Target power")
  
  if (print) return(invisible(res)) else return(res)
}
